### PR TITLE
Sean+kyle/gfc 706/hmrc tax period4

### PR DESCRIPTION
**Pull request by kylemchownhmrc**

### CODE REVIEW

File: app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala [hunk 1]
```diff
@@ -476,8 +510,54 @@ class SectionRenderingService(
...
     }
 
+  private def htmlForHmrcTaxPeriod(
+    fieldValue: FormComponent,
+    index: Int,
+    ei: ExtraInfo,
+    validatedType: ValidatedType,
+    data: FormDataRecalculated,
+    obligations: Map[HmrcTaxPeriod, TaxPeriods],
+    HmrcTP: HmrcTaxPeriod) = {
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala]: Capitalisation: needs to be hmrcTP

File: app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala [hunk 2]
```diff
@@ -476,8 +510,54 @@ class SectionRenderingService(
...
+            formatDate(i.inboundCorrespondenceFromDate),
+            formatDate(i.inboundCorrespondenceToDate),
+            false))
+    val validatedValue = buildFormFieldValidationResult(fieldValue, ei, validatedType, data)
+    val mapOfResultsOption = validatedValue match {
+      case Some(ComponentField(a, b)) => b
+      case _                          => Map[String, FormFieldValidationResult]()
+    }
+    val setValue = mapOfResultsOption.values.toList match {
+      case a => TextFormatter.formatText(Some(a(0))).takeRight(4)
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala]: Why 4?

File: app/uk/gov/hmrc/gform/models/helpers/TaxPeriodHelper.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.gform.models.helpers
+import java.text.SimpleDateFormat
+import java.util.Date
+
+import uk.gov.hmrc.gform.sharedmodel.formtemplate.{ FileUpload, FormComponent, FormComponentId }
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/models/helpers/TaxPeriodHelper.scala]: Not required

File: app/uk/gov/hmrc/gform/obligation/ObligationService.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+package uk.gov.hmrc.gform.obligation
+import java.text.SimpleDateFormat
+
+import uk.gov.hmrc.gform.gformbackend.GformConnector
+import uk.gov.hmrc.gform.sharedmodel._
+import uk.gov.hmrc.gform.sharedmodel.formtemplate._
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.{ ExecutionContext, Future }
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/obligation/ObligationService.scala]: Future not needed

File: app/uk/gov/hmrc/gform/sharedmodel/TaxPeriods.scala [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+object Identification {
+  implicit val format: OFormat[Identification] = Json.format[Identification]
+}
+
+case class ObligationDetail(
+  status: String,
+  inboundCorrespondenceFromDate: String,
+  inboundCorrespondenceToDate: String,
+  inboundCorrespondenceDueDate: String,
```
Comment (lutnos) [inline app/uk/gov/hmrc/gform/sharedmodel/TaxPeriods.scala]: ~These three should be of Date type, converted from strings in the gform backend

File: app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala [hunk 6]
```diff
@@ -476,8 +510,54 @@ class SectionRenderingService(
...
+            formatDate(i.inboundCorrespondenceFromDate),
+            formatDate(i.inboundCorrespondenceToDate),
+            false))
+    val validatedValue = buildFormFieldValidationResult(fieldValue, ei, validatedType, data)
+    val mapOfResultsOption = validatedValue match {
+      case Some(ComponentField(a, b)) => b
+      case _                          => Map[String, FormFieldValidationResult]()
+    }
+    val setValue = mapOfResultsOption.values.toList match {
+      case a => TextFormatter.formatText(Some(a(0))).split("|")(2)
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/gform/SectionRenderingService.scala]: Access to `a(0)` is unsafe. Why don't use `case a :: _ => TextFormatter.formatText(Some(a)).split("|")(2)` instead ?

File: app/uk/gov/hmrc/gform/views/summary/snippets/hmrc_tax_period.scala.html [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+
+@hasErrors = @{validationResult.exists(_.isNotOk)}
+
+@errors = @{validationResult.map(_.fieldErrors.toList).getOrElse(Set().toList).map{ e =>
+    errorInline("summary", e, Seq()) }}
+
+@formatTaxPeriod(taxPeriod:String) = @{
+val l =taxPeriod.split('|')
+localisation("From")+" "+l(0)+" "+localisation("to")+" "+l(1)
```
Comment (josef-vlach) [inline app/uk/gov/hmrc/gform/views/summary/snippets/hmrc_tax_period.scala.html]: Parsing taxPeriod is unsafe. Why don't do something like this:

```
taxPeriod.split('|').toList match {
  case from :: to :: _ => localisation("From")+" "+from+" "+localisation("to")+" "+to
  case _ => ??? // Too late to react on missing pipe
}
```

But it still isn't perfect as we cannot react on wrong data. This should be done much earlier and `OptionParams` (I guess) object should be pass here.

**Pull request merged**